cmd/internal/backup/providers/contract.go

```diff
@@ -10,8 +11,8 @@ type BackupProvider interface {
 	ListBackups(ctx context.Context) (BackupVersions, error)
 	CleanupBackups(ctx context.Context) error
 	GetNextBackupName(ctx context.Context) string
-	DownloadBackup(ctx context.Context, version *BackupVersion, outDir string) (string, error)
-	UploadBackup(ctx context.Context, sourcePath string) error
+	DownloadBackup(ctx context.Context, version *BackupVersion, writer io.Writer) error
+	UploadBackup(ctx context.Context, reader io.Reader, sourcePath string) error
```

Comment:
We should try to eliminate `sourcePath` here when we pass the reader. Let's introduce static names for a backup provider and initialize them with the knowledge if content is encrypted or not in order to deal with the `.aes` file extension.

Reply:
what do you mean by static names for a backup-provider. Like a flag for the backup-provider if encryption is enabled or not?

This solution would work now because we just use streams for the backup-provider. If we add streams to encrypter we may have problems with the transition: 

This may result in following result:
1. old-backups are not encrypted
2. backup-restore-sidecar gets redeployed with encryption
  1. if doesn't need to restore -> no problem
  2. if needs to restore -> tries do decrypt not encrypted backup
  
This is just an edge case. But am I wrong here?

Reply:
Sorry, I was not fully clear. From the `sourcePath` argument we now only derive the file name that is used inside the backup bucket (`destination`).

We can build the destination file name already at the startup of the program because it remains constant. So, when initializing a backup provider, we could pass the compressor and the encrypter. The compressor already provides an `Extension()` function, which would be helpful for the encrypter, too. Every backup provider could have constant, which would be `db` if we do not want to break anything.

So, the resulting file name for a backup would be:

```golang
backupFileName := strings.Join([]string{ProviderConstant, comp.Extension(), encrypter.Extension()}, ".")
```

This refactoring does not change any of the current behavior, so I don't think that the edge-case scenario is any different from what it is now.